### PR TITLE
Restructure wrapper to fix memory management

### LINKED ISSUE #24: Segmentation fault while solving basis pursuit problem with convex.jl
I have the following code for a basis pursuit simulation:

``` julia
using Convex
import SCS.SCSSolver
solver = SCSSolver(verbose=0);

function sparse_recovery(matrix_gen; n=100, k_range=1:20, trials=100)
    for k = k_range
        for m = [[1:4]; [5:5:n]]
            recoveries = 0.
            for i = 1:trials
                A = matrix_gen(m, n)
                x0 = sprandn(n, 1, k / n)
                y = A * x0

                x = Variable(n)
                problem = minimize(norm(x, 1), [y == A * x])
                solve!(problem, solver)

                # perfect recovery is infinity norm < 0.001
                recoveries += (norm(x.value - x0, Inf) < 0.001)
            end

            success_rate = recoveries / trials

            if success_rate <= 0.05 || success_rate >= 0.95
                @printf("%d-sparse, %d measures: %.2f recovery\n", k, m, success_rate)
            end

            if success_rate >= 0.95
                break
            end
        end
        println()
    end
end

sparse_recovery(randn)
```

Running this works for a while (several thousand calls to `solve!`) but eventually seg-faults:

```
signal (11): Segmentation fault
unNormalizeA at /home/emartin/.julia/v0.3/SCS/deps/src/scs-1.0.7/linsys/common.c:198
scs_finish at /home/emartin/.julia/v0.3/SCS/deps/src/scs-1.0.7/src/scs.c:706
SCS_finish at /home/emartin/.julia/v0.3/SCS/src/low_level_wrapper.jl:42
jlcall_SCS_finish_21851 at  (unknown line)
jl_apply_generic at /usr/bin/../lib/x86_64-linux-gnu/julia/libjulia.so (unknown line)
SCS_solve at /home/emartin/.julia/v0.3/SCS/src/high_level_wrapper.jl:131
julia_SCS_solve_21832 at  (unknown line)
jlcall_SCS_solve_21832 at  (unknown line)
optimize! at /home/emartin/.julia/v0.3/SCS/src/SCSSolverInterface.jl:145
jlcall_optimize!_21830 at  (unknown line)
jl_apply_generic at /usr/bin/../lib/x86_64-linux-gnu/julia/libjulia.so (unknown line)
solve! at /home/emartin/.julia/v0.3/Convex/src/solution.jl:58
julia_solve!_21465 at  (unknown line)
jl_apply_generic at /usr/bin/../lib/x86_64-linux-gnu/julia/libjulia.so (unknown line)
sparse_recovery at none:16
julia_sparse_recovery_21384 at  (unknown line)
jlcall_sparse_recovery_21384 at  (unknown line)
jl_apply_generic at /usr/bin/../lib/x86_64-linux-gnu/julia/libjulia.so (unknown line)
unknown function (ip: -1343619028)
unknown function (ip: -1343624175)
unknown function (ip: -1343562680)
jl_f_top_eval at /usr/bin/../lib/x86_64-linux-gnu/julia/libjulia.so (unknown line)
eval_user_input at REPL.jl:53
jlcall_eval_user_input_19972 at  (unknown line)
jl_apply_generic at /usr/bin/../lib/x86_64-linux-gnu/julia/libjulia.so (unknown line)
anonymous at task.jl:95
jl_handle_stack_switch at /usr/bin/../lib/x86_64-linux-gnu/julia/libjulia.so (unknown line)
julia_trampoline at /usr/bin/../lib/x86_64-linux-gnu/julia/libjulia.so (unknown line)
unknown function (ip: 4199453)
__libc_start_main at /lib/x86_64-linux-gnu/libc.so.6 (unknown line)
unknown function (ip: 4199509)
zsh: segmentation fault (core dumped)  julia
```

I'm not sure if the bug lies in Convex.jl, SCS.jl, or SCS. I'm debugging now with gdb, will update issue as I learn more.

### LINKED ISSUE #74: ERROR: A->p (column pointers) decreasing
I've seen a much older bug with similar message.
This happens on julia-0.5.1 with openblas and SCS/master 
```
SDP_problem = Maximization problem with:
 * 2018882 linear constraints
 * 1 semidefinite constraint
 * 1959211 variables
Solver is default solver
solver = SCS.SCSSolver(Any[(:eps,1.0e-8),(:max_iters,100000),(:verbose,true)])
ERROR: A->p (column pointers) decreasing
invalid linear system input data
ERROR: Validation returned failure
ERROR: SCS_NULL input
```
This is how I generate the problem:
```
function create_SDP_problem(matrix_constraints, Δ, Δ², N)
    @assert length(Δ) == length(matrix_constraints)
    m = JuMP.Model();
    JuMP.@variable(m, A[1:N, 1:N], SDP)
    JuMP.@SDconstraint(m, A >= 0)
    JuMP.@constraint(m, sum(A[i] for i in eachindex(A)) == 0)
    JuMP.@variable(m, κ >= 0.0)
    JuMP.@objective(m, Max, κ)

    for (pairs, δ², δ) in zip(matrix_constraints, Δ².coefficients, Δ.coefficients)
        JuMP.@constraint(m, sum(A[i,j] for (i,j) in pairs) == δ² - κ*δ)
    end
    return m
end
```
I haven't seen this on a much smaller problem generated by the exact same code:
```
SDP_problem = Maximization problem with:
 * 157204 linear constraints
 * 1 semidefinite constraint
 * 156521 variables
```

### LINKED ISSUE #90: critical memory bug
If I'm understanding things correctly [`create_scs_matrix`](https://github.com/JuliaOpt/SCS.jl/blob/d63bc8ca6a8d56ee1ec046589c1f211751482f9a/src/high_level_wrapper.jl#L5) returns pointers to vectors that can be freed by the Julia GC at any point. I'm surprised we haven't seen more segfaults.

@bodono, does SCS make an internal copy of the `A` matrix, or does it assume that the user won't free it before `SCS_solve`?

### PR DESCRIPTION
**Pull request by mlubin**
Breaking changes to interface of SCS_solve.

Closes #90 
Probably Closes #24 
Probably Closes #74 

CC @blegat (I removed your deprecations)

### CODE REVIEW

File: src/types.jl [hunk 1]
```diff
@@ -12,18 +11,48 @@ struct SCSMatrix
...
+
 
 struct SCSSettings
     normalize::Int # boolean, heuristic data rescaling: 1
     scale::Cdouble # if normalized, rescales by this factor: 1
     rho_x::Cdouble # x equality constraint scaling: 1e-3
-    max_iters::Int # maximum iterations to take: 2500
-    eps::Cdouble # convergence tolerance: 1e-3
+    max_iters::Int # maximum iterations to take: 5000
+    eps::Cdouble # convergence tolerance: 1e-5
```
Comment (blegat) [inline src/types.jl]: Is this related to #59  ? The value does not match what I say in the issue. Has it changed with the latest version of SCS ? The README should probably also be updated

Comment (mlubin) [reply src/types.jl]: The defaults have changed, but it's not a good idea to have them listed in three places. I removed them from the README.

**Comment by mlubin** [comment]
This PR closes 50% of all open issues on SCS.jl. Just need to figure out why it's crashing on 0.7.

**Comment by mlubin** [comment]
The crash on 0.7 was related to https://discourse.julialang.org/t/isbits-object-lifetimes-on-0-7/7482

**Comment by mlubin** [comment]
This is passing locally, the travis failure on linux nightly is due to https://github.com/JuliaLang/BinDeps.jl/issues/343.

**Pull request merged**